lib/odata_duty.rb

```diff
@@ -253,91 +218,16 @@ def self.index_hash(metadata_url)
-      match_data = url.match(/.*\(([^)]+)\)/)
-      return nil unless match_data
-
-      value = match_data[1]
-
-      value.gsub(/^"|"$/, '').gsub(/^'|'$/, '')
-    end
-
-    require 'delegate'
-    class ContextWrapper < SimpleDelegator
-      attr_accessor :endpoint
-    end
-
     def self.execute(url, context:, query_options: {})
-      endpoint = points.find { |e| url.split('/$count').first.split('(').first == e.url }
-      raise UnknownPropertyError, "No endpoint #{url} found in #{urls}" unless endpoint
-
-      context = ContextWrapper.new(context)
-      context.endpoint = endpoint
-
-      entity_id = extract_value_from_brackets(url)
-      return individual(endpoint, entity_id, context) if url.include?('(')
-
-      set_builder = prepare_builder(endpoint, context, query_options)
-      return set_builder.count if url.include?('/$count')
-
-      collection(set_builder, endpoint, context, query_options)
-    end
-
-    def self.prepare_builder(endpoint, context, query_options)
-      endpoint.new_entity_set(context: context).tap do |set_builder|
-        if query_options.key?('$filter')
-          endpoint.apply_filter(set_builder, query_options['$filter'])
-        end
-      end
-    end
-
-    require 'oj'
-
-    def self.individual(endpoint, entity_id, context)
-      Oj.dump(
-        endpoint
-          .individual(entity_id, context: context)
-          .merge(
-            '@odata.context': context.url_for(url: '$metadata', anchor: "#{endpoint.name}/$entity")
-          ),
-        mode: :compat
-      )
-    end
-
-    def self.collection(set_builder, endpoint, context, query_options)
-      count = set_builder.count if query_options['$count'] == 'true'
-      apply_remaining(query_options, endpoint, set_builder)
-      data = endpoint
-             .collection(set_builder, context: context)
-             .merge('@odata.context': context.url_for(url: '$metadata', anchor: endpoint.name))
-      data['@odata.count'] = count if count
-      add_next_link(data, endpoint, set_builder, query_options, context)
-      Oj.dump(data, mode: :compat)
-    end
-
-    def self.apply_remaining(query_options, endpoint, set_builder)
-      query_options.except('$count', '$filter')
-                   .select { |k, _| k.start_with?('$') }.each do |k, v|
-        endpoint.public_send("apply_#{k[1, 10]}", set_builder, v)
-      rescue NoMethodError
-        raise NoImplementionError, "query option #{k} not supported"
-      end
-    end
-
-    def self.add_next_link(data, endpoint, set_builder, query_options, context)
-      return unless set_builder.od_next_link_skiptoken
-
-      next_query_options = query_options.merge(url: endpoint.url,
-                                               :$skiptoken => set_builder.od_next_link_skiptoken)
-      data[:'@odata.nextLink'] = context.url_for(**next_query_options)
+      Executor.execute(url: url, context: context, query_options: query_options, schema: self)
```

Comment:
moved a lot of the logic into Executor class, so the same execution logic can work for static and dynamic schemas.